optuna/_callbacks.py

```diff
@@ -0,0 +1,56 @@
+from typing import Tuple
+
+from optuna._experimental import experimental
+from optuna.trial import FrozenTrial
+from optuna.trial import TrialState
+
+
+@experimental("2.8.0")
+class MaxTrialsCallback:
+    """Set a maximum number of trials before ending the study.
+
+    While the :obj:`n_trials` argument of :obj:`optuna.optimize` sets the number of trials that
+    will be run, you may want to continue running until you have a certain number of successfullly
+    completed trials or stop the study when you have a certain number of trials that fail.
+    This :obj:`MaxTrialsCallback` class allows you to set a maximum number of trials for a
+    particular :class:`~optuna.trial.TrialState` before stopping the study.
+
+    Example:
+
+        .. testcode::
+
+            import optuna
+            from optuna.study import MaxTrialsCallback
+            from optuna.trial import TrialState
+
+
+            def objective(trial):
+                x = trial.suggest_float("x", -1, 1)
+                return x ** 2
+
+
+            study = optuna.create_study()
+            study.optimize(
+                objective,
+                callbacks=[MaxTrialsCallback(10, states=(TrialState.COMPLETE,))],
+            )
+
+    Args:
+        n_trials:
+            The max number of trials. Must be set to an integer.
+        states:
+            Tuple of the :class:`~optuna.trial.TrialState` to be counted
+            towards the max trials limit. Default value is :obj:`(TrialState.COMPLETE,)`.
+    """
+
+    def __init__(
+        self, n_trials: int, states: Tuple[TrialState, ...] = (TrialState.COMPLETE,)
+    ) -> None:
+        self._n_trials = n_trials
+        self._states = states
+
+    def __call__(self, study: "optuna.study.Study", trial: FrozenTrial) -> None:  # type: ignore # noqa
```

Comment:
More of a question but how come we need to ignore the type/style checks? mypy is for instance skipped now. I'm wondering if you cannot `import optuna` without circular imports, and thereby "fix" the type hint?

Reply:
My local environment has results different from CI 🤷‍♂️ , so will test again with CI.

Reply:
`# NOQA` required for flake8: https://github.com/optuna/optuna/pull/2612/checks?check_run_id=2572273720

Reply:
Confirmed `type: ignore` required for mypy test: https://github.com/optuna/optuna/pull/2612/checks?check_run_id=2572273720

Reply:
Error when using `import optuna` instead of `study: "optuna.study.Study"`:

```
➜  Optuna git:(max_callbacks) ✗ python examples/max_trials_callback.py 
Traceback (most recent call last):
  File "examples/max_trials_callback.py", line 14, in <module>
    import optuna
  File "/Users/Crissman/GDrive/Github/optuna/optuna/__init__.py", line 8, in <module>
    from optuna import importance
  File "/Users/Crissman/GDrive/Github/optuna/optuna/importance/__init__.py", line 6, in <module>
    from optuna.importance._base import BaseImportanceEvaluator
  File "/Users/Crissman/GDrive/Github/optuna/optuna/importance/_base.py", line 9, in <module>
    from optuna.samplers import intersection_search_space
  File "/Users/Crissman/GDrive/Github/optuna/optuna/samplers/__init__.py", line 1, in <module>
    from optuna.samplers._base import BaseSampler
  File "/Users/Crissman/GDrive/Github/optuna/optuna/samplers/_base.py", line 8, in <module>
    from optuna.study import Study
  File "/Users/Crissman/GDrive/Github/optuna/optuna/study.py", line 21, in <module>
    from optuna._callbacks import MaxTrialsCallback  # NOQA
  File "/Users/Crissman/GDrive/Github/optuna/optuna/_callbacks.py", line 9, in <module>
    @experimental("2.8.0")
  File "/Users/Crissman/GDrive/Github/optuna/optuna/_callbacks.py", line 53, in MaxTrialsCallback
    def __call__(self, study: optuna.study.Study, trial: FrozenTrial) -> None:
AttributeError: module 'optuna' has no attribute 'study'
```

Reply:
This seems to work https://github.com/hvy/optuna/commit/546cbe7d00bef14216b87430b40edf19a983f09f.

Reply:
Aha. Not *instead of* but *and*. Ok, did that. 👍

Reply:
Sorry, should've been more clear. Thanks!